### PR TITLE
chore(deps): Upgrade to Django 4.1

### PR DESCRIPTION
**Pull request by webjunkie**
## Problem

Upgrade to a more recent Django version.

## Changes

- upgrading to Django 4.1 seems straightforward
- further upgrading to Django 4.2 poses some issues: see #18653 to track these

## How did you test this code?

- look at CI

### CODE REVIEW

File: posthog/api/test/dashboards/__snapshots__/test_dashboard.ambr [hunk 1]
```diff
@@ -954,6 +954,9 @@
               OR "posthog_dashboardtile"."insight_id" IS NULL)
          AND "posthog_dashboardtile"."dashboard_id" = 2
          AND "posthog_dashboardtile"."dashboard_id" = 2
+         AND NOT ("posthog_dashboard"."deleted"
+                  AND "posthog_dashboardtile"."deleted"
+                  AND "posthog_dashboardtile"."deleted" IS NOT NULL)
```
Comment (pauldambra) [inline posthog/api/test/dashboards/__snapshots__/test_dashboard.ambr:959]: interesting change... I guess it's only enforcing that we don't return deleted things so probably fine

**Review by pauldambra — APPROVED** [review]
clicking around locally things seem to work 

(Web analytics doesn't load anything (but it doesn't in master for me now and I'd not loaded it locally before)

I think the roll out is the interesting bit. Whether we merge with folk on standby for a fast rollback if things break in production unexpectedly or if we try and figure out how to get it into dev separately to deploying to production

**Comment by pauldambra** [comment]
i added the deploy label - pushing a new change deploys an ephemeral environment we could use as another functionality check

**Comment by pauldambra** [comment]
I think I've pushed a commit that fixes that 🤞

**Comment by webjunkie** [comment]
Seems it now ran through but the deploy is not accessible 🤔

**Comment by pauldambra** [comment]
Screenshot 2023-11-30 at 09 42 15
yeah, it seems unhappy.... I managed to get to login but not beyond that. Can't get logs from pods

@frankh is this something we've done or the deploy environment itself (or how should we check 😊)

**Comment by webjunkie** [comment]
Found one issue with CSRF thanks due to hobby deploy 👏🏻 

https://docs.djangoproject.com/en/4.2/releases/4.0/#csrf-trusted-origins-changes

File: posthog/settings/access.py [hunk 2]
```diff
@@ -25,6 +25,10 @@
 if get_from_env("DISABLE_SECURE_SSL_REDIRECT", False, type_cast=str_to_bool):
     SECURE_SSL_REDIRECT = False
 
+CSRF_TRUSTED_ORIGINS = [
+    "https://*.posthog.com",
+    "https://*.posthog.dev",
```
Comment (webjunkie) [inline posthog/settings/access.py:30]: Setting this was required to make some requests work in the preview environment.

What could we set this to in general? Is there some hostname already flying around?

Comment (pauldambra) [reply posthog/settings/access.py:30]: Screenshot 2023-12-08 at 13 08 12

Ah, missed that scanning the files 🤦 

https://github.com/PostHog/posthog/blob/master/posthog/settings/__init__.py#L56

Comment (webjunkie) [reply posthog/settings/access.py:30]: That would fit well, will try.

**Comment by webjunkie** [comment]
I can log in and things seem to work. I've just fixed some type problems so then CI should be green.

What else can we check or how do we proceed?

**Comment by pauldambra** [comment]
I'm not sure forcing the image for a deploy to dev gives us much more than the deploy label does.

@frankh is infrastructure support this week. So maybe we make sure we've got his time booked and deploy this with him on hand to force a rollback if we see issues?

(or we ask him how we can force an image in dev to try this there...)

**Review by danielxnj — APPROVED** [review]
LGTM

**Pull request merged**